src/SimplifyCorrelatedDifferences.cpp

```diff
@@ -45,14 +51,33 @@ class SimplifyCorrelatedDifferences : public IRMutator {
         std::vector<Frame> frames;
         StmtOrExpr result;
 
+        // Note that we must add *everything* that depends on the loop
+        // var to the monotonic scope and the list of lets, even
+        // things which we can never substitute in (e.g. impure
+        // things). This is for two reasons. First this pass could be
+        // used at a time when we still have nested lets under the
+        // same name. If we decide not to add an inner let, but do add
+        // the outer one, then later references to it will be
+        // incorrect. Second, if we don't add something that happens
+        // to be non-monotonic, then is_monotonic finds a variable
+        // that references it in a later let, it will think it's a
+        // constant, not an unknown.
         do {
             result = op->body;
-            if (op->value.type() == Int(32) && is_pure(op->value)) {
+            if (loop_var.empty()) {
+                frames.emplace_back(op);
+                continue;
+            }
+
+            bool pure = is_pure(op->value);
+            if (!pure || expr_uses_vars(op->value, monotonic)) {
                 frames.emplace_back(op, loop_var, monotonic);
                 Expr new_value = mutate(op->value);
-                lets.emplace_back(op->name, new_value);
+                bool may_substitute_in = new_value.type() == Int(32) && pure;
+                lets.emplace_back(OuterLet{op->name, new_value, may_substitute_in});
```

Comment:
for emplace_back, shouldn't it just be emplace_back(op->name, new_value, may_substitute_in)?